src/pages/content-edit/head.js

```diff
@@ -3,261 +3,431 @@ import format from "date-fns/format";
+
+        updated = {
+            unpublished_at : nowTs,
+            unpublished_by : user
         };
+
+        if(ctrl.start.ts &&  ctrl.start.ts > ctrl.end.ts) {
+            // Invalid start.
+            ctrl.start = ctrl.nulledScheduleStart();
+        }
+
+        ref.update(updated, function() {
+            ctrl.saving = false;
+            m.redraw();
+        });
+    };
+
+    ctrl.clearSchedule = function() {
+        ctrl.start = ctrl.nulledScheduleStart();
+        ctrl.end   = ctrl.nulledScheduleEnd();
+        ctrl.invalidDates = false;
     };
 
     ctrl.save = function(opts) {
-        ctrl.saving = true;
+        var updated = {};
 
+        ctrl.saving = true;
         m.redraw();
 
-        ref.update({
+        updated = {
             fields : opts.data.fields,
             name   : opts.data.name,
             slug   : opts.data.slug || null
-        }, function() {
-            ctrl.saving = false;
+        };
+
+        updated = ctrl.addScheduleData(updated);
 
+        ref.update(updated, function() {
+            ctrl.saving = false;
             m.redraw();
         });
     };
+
+    ctrl.addScheduleData = function(updated) {
+        if(!ctrl.invalidDates) {
+            let startTs = null,
+                endTs = null,
+                startTime = ctrl.start.time || DEFAULT_START_TIME,
+                endTime = ctrl.end.time || DEFAULT_END_TIME;
+
+            if(ctrl.start.date) {
+                startTs = ctrl.getTimestampFromStr(ctrl.start.date + " " + startTime);
+            }
+            if(ctrl.end.date) {
+                endTs = ctrl.getTimestampFromStr(ctrl.end.date + " " + endTime);
+            }
+
+            updated.published_at = startTs;
+            updated.unpublished_at = endTs;
+
+            if(options.data.published_at && !updated.published_at) {
+                // Publish date was removed. Null it out.
+                updated.published    = null;
+                updated.published_at = null;
+                updated.published_by = null;
+
+                ctrl.start = ctrl.nulledScheduleStart();
+            }
+
+            if(options.data.unpublished_at && !updated.unpublished_at) {
+                // Unpublish date was removed. Null it out.
+                updated.unpublished_at = null;
+                updated.unpublished_by = null;
+
+                ctrl.end = ctrl.nulledScheduleEnd();
+            }
+        }
+
+        return updated;
+    };
+
+
+    ctrl.recalculateTimestamps = function() {
+        var start = ctrl.start,
+            end = ctrl.end;
+
+        if(start.date) {
+            start.ts = ctrl.getTimestampFromStr(start.date + " " + start.time);
+        }
+        if(end.date) {
+            end.ts = ctrl.getTimestampFromStr(end.date + " " + end.time);
+        }
+
+        if(start.date && end.date) {
+            ctrl.invalidDates = (start.ts > end.ts);
+        }
+    };
+
+    ctrl.makeScheduleObj = function(ts) {
```

Comment:
What would you prefer? I'd rather not add floating methods into the controller...  constructor? Is this technically a constructor? Anyway -- I think it muddies the readability because you must declare them at the top of the scope, so you end up with all these floaty utility functions you're scanning through without any context about how they're being used before you get to the start of anyl code path.

Would you like me to put them outside of the `controller() { ... }` context? That feels odd because its functionality is limited to the controller scope.  Or would you like me to declare these utilities at the top of the controller scope?